### DISCUSSION TITLE
Masterkey Rotation for all the nodes.

### DISCUSSION THREAD

**Discussion by shubhanshugupta96** [openthread/openthread#9105] [Q&A] [answered]
Hi, 
I have an existing network of thread devices. 
SEDs and FTDs are deployed in various corners of the building and border router is raspberry pi with NCP.
I want to rotate the masterkey without reprogramming my nodes. I am fine by making some changes in Raspberry Pi.
I tried to implement a method by giving the security policy as  " 1 onrc" in active dataset by changing my implementation to RCP instead of NCP. But it did not change the master key after 1 hr.
Is there a possibility I can do this. 

Attached active dataset :
Active Timestamp: 1
Channel: 25
Channel Mask: 0x07fff800
Ext PAN ID: 73fb75d4d448e284
Mesh Local Prefix: fdfd:5e4:d8ff:f57f::/64
Network Key: 73fb75d4d448e284315172d9f0f42b28
Network Name: testnetwork
PAN ID: 0x73fb
PSKc: 7453e23467cff92b384d1ca3fb7861b2
Security Policy: 1 onrc

This was same before 1hr and after.

Thanks and Regards,
Shubhanshu Gupta

**Comment by jwhui** [comment, marked as answer]
Best practice when updating the Network Key is to set a new Pending Operational Dataset.

- dataset init active
- dataset networkkey <key>
- dataset pendingtimestamp 1
- dataset delay 30000
- dataset commit pending

After 30 seconds, the Pending Operational Dataset should be transferred to the Active Operational Dataset.

**Reply by shubhanshugupta96** [reply]
Thanks a lot for a quick response, that worked.
I have a doubt, since my FTDs are wall powered, there is a possibility that one or more of my device is switched off at the time of sharing the pending dataset and they gets switched on after the dataset is changed. If that happens, my network will get disturbed as the device will not have an active dataset and will not be able to send data in the network.

**Reply by jwhui** [reply]
Yes, that is a risk of changing the network key.

**Reply by shubhanshugupta96** [reply]
Don't we have a solution for this?

**Reply by jwhui** [reply]
Thread does not provide an alternative today. Thread security currently relies on a shared network key. Changing the network key implies that the old network key should no longer be used. If you'd like to propose updates to the Thread Specification, I encourage you to join the [Thread Group](https://www.threadgroup.org/) and contribute.
